File: reference/python/autopilot/controller.py
"""CourseKeeper: the top-level "hold a true course" loop.

It owns the flow the user described:

    dial in a heading  ->  read NMEA heading  ->  compare to desired
                       ->  PID (damped) rudder command  ->  drive the ram

and adds the operational glue: engage/standby, deriving yaw rate when the
sensor doesn't supply one, speed-scheduled gains, and a low-steerage guard so
the pilot doesn't thrash the helm when the boat is barely moving.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Tuple

from .config import AutopilotConfig
from .heading import heading_error, normalize_180, normalize_360, yaw_rate as derive_yaw_rate
from .pid import PID
from .ram import RamActuator


@dataclass
class CourseKeeperConfig:
    """Thin wrapper so the controller can be built from an AutopilotConfig."""

    autopilot: AutopilotConfig = field(default_factory=AutopilotConfig)


@dataclass
class ControlTick:
    """Diagnostics returned from each :meth:`CourseKeeper.update` call."""

    engaged: bool
    desired_heading: float
    measured_heading: float
    error: float
    yaw_rate: float
    rudder_command: float
    rudder_achieved: float
    speed_knots: float
    reason: str = ""


class CourseKeeper:
    """Closed-loop heading-hold controller.

    Wire it to a :class:`~autopilot.ram.RamActuator` (SimRam for testing, a
    HardwareRam on the boat). Feed :meth:`update` a fresh heading each control
    tick; it returns a :class:`ControlTick` for logging/telemetry.
    """

    def __init__(self, config: AutopilotConfig, ram: RamActuator):
        self.config = config
        self.ram = ram
        self.pid = PID(config.pid)
        self._engaged = False
        self._desired: Optional[float] = None
        self._prev_heading: Optional[float] = None
        self._filtered_heading: Optional[float] = None
        self._base_gains = config.pid.gains

    # ---- engage / setpoint --------------------------------------------
    def engage(self, desired_heading: Optional[float] = None,
               current_heading: Optional[float] = None) -> None:
        """Engage the pilot.

        With no ``desired_heading`` it locks onto the current heading - the
        common "press AUTO and hold what I've got" case.
        """
        if desired_heading is None:
            desired_heading = current_heading
        if desired_heading is None:
            raise ValueError("engage needs a desired or current heading")
        self._desired = normalize_360(desired_heading)
        self._engaged = True
        self.pid.reset()
        self._prev_heading = current_heading
        self._filtered_heading = (normalize_360(current_heading)
                                  if current_heading is not None else None)

    def standby(self) -> None:
        """Disengage: stop driving the ram and let the helm go free."""
        self._engaged = False
        self.ram.stop()

    def set_heading(self, desired_heading: float) -> None:
        """Dial in a new course to hold (deg true)."""
        self._desired = normalize_360(desired_heading)

    def adjust_heading(self, delta_deg: float) -> None:
        """Nudge the target, e.g. +1 / -1 / +10 buttons on the pilot head."""
        if self._desired is None:
            raise ValueError("no heading set")
        self._desired = normalize_360(self._desired + delta_deg)

    @property
    def engaged(self) -> bool:
        return self._engaged

    @property
    def desired_heading(self) -> Optional[float]:
        return self._desired

    # ---- speed-scheduled gains ----------------------------------------
    def _apply_speed_scheduling(self, speed_knots: float) -> None:
        """Re-tune the PID for the current speed from the Nomoto model.

        Rudder authority scales ~speed^2; without this the pilot hunts at hull
        speed and goes limp when slow. We re-derive gains keeping the same
        damping ratio, so smoothness is preserved across the speed range.
        """
        if not self.config.speed_scheduling:
            self.pid.config.gains = self._base_gains
            return
        v = max(speed_knots, self.config.min_speed_knots)
        scaled = self.config.nomoto.scaled_to_speed(v, self.config.ref_speed_knots)
        try:
            # Re-derive gains for the scaled plant using the SAME feel targets
            # (damping ratio + bandwidth) as the base tuning, so smoothness is
            # preserved across the speed range instead of silently changing.
            self.pid.config.gains = self.config.gains_for(scaled)
        except ValueError:
            self.pid.config.gains = self._base_gains

    # ---- main loop -----------------------------------------------------
    def update(
        self,
        measured_heading: float,
        dt: float,
        sensor_yaw_rate: Optional[float] = None,
        speed_knots: Optional[float] = None,
    ) -> ControlTick:
        """Run one control tick.

        :param measured_heading: heading from NMEA (deg).
        :param dt: seconds since the previous tick.
        :param sensor_yaw_rate: yaw rate from a ROT sentence / gyro (deg/s).
            If ``None`` it's derived from successive headings.
        :param speed_knots: boat speed for gain scheduling / steerage guard.
        """
        raw = normalize_360(measured_heading)
        speed = speed_knots if speed_knots is not None else self.config.ref_speed_knots

        # Low-pass the compass input (wrap-aware). A raw compass is noisy, and
        # differentiating noise for the counter-rudder term is the fastest route
        # to a buzzing helm. Real fluxgate/GPS compasses are physically damped;
        # this models that so the derivative term sees a clean signal.
        tau = self.config.heading_filter_tau
        if tau > 0.0 and self._filtered_heading is not None:
            alpha = dt / (tau + dt)
            self._filtered_heading = normalize_360(
                self._filtered_heading + alpha * normalize_180(raw - self._filtered_heading))
        else:
            self._filtered_heading = raw
        measured = self._filtered_heading

        # Yaw rate: prefer a real sensor (clean), else differentiate the
        # *filtered* heading.
        if sensor_yaw_rate is not None:
            rate = sensor_yaw_rate
        elif self._prev_heading is not None:
            rate = derive_yaw_rate(self._prev_heading, measured, dt)
        else:
            rate = 0.0
        self._prev_heading = measured

        if not self._engaged or self._desired is None:
            return ControlTick(
                engaged=False, desired_heading=self._desired or 0.0,
                measured_heading=measured, error=0.0, yaw_rate=rate,
                rudder_command=self.ram.rudder_angle,
                rudder_achieved=self.ram.rudder_angle,
                speed_knots=speed, reason="standby",
            )

        error = heading_error(self._desired, measured)

        # Low-steerage guard: with no water flow over the rudder the helm has no
        # authority, so hold station rather than winding up and slamming over.
        if speed_knots is not None and speed_knots < self.config.min_speed_knots:
            achieved = self.ram.command(self.ram.rudder_angle, dt)
            return ControlTick(
                engaged=True, desired_heading=self._desired, measured_heading=measured,
                error=error, yaw_rate=rate, rudder_command=self.ram.rudder_angle,
                rudder_achieved=achieved, speed_knots=speed, reason="low-steerage",
            )

        self._apply_speed_scheduling(speed)
        rudder_cmd = self.pid.update(error, rate, dt)
        achieved = self.ram.command(rudder_cmd, dt)

        return ControlTick(
            engaged=True, desired_heading=self._desired, measured_heading=measured,
            error=error, yaw_rate=rate, rudder_command=rudder_cmd,
            rudder_achieved=achieved, speed_knots=speed, reason="hold",
        )

File: internal/controller/controller.go
// Package controller is the top-level "hold a course" loop.
//
// It supports two modes:
//
//   - HeadingHold: hold a compass heading. error = desired - measured, fed to
//     the damped PID which drives the ram.
//
//   - TrackCOG: hold a course over ground. This is a cascade. The heading-hold
//     PID stays as the fast inner loop; a slow outer loop nudges the target
//     heading so the *actual* COG matches the desired ground course. The offset
//     the outer loop settles on is exactly the crab angle needed to counter
//     current and leeway. COG is noisy and meaningless at low speed, so below
//     MinTrackSpeed the controller falls back to holding heading. Time-scale
//     separation (inner ~8 s, outer ~40 s) keeps the cascade stable.
package controller

import (
	"math"

	"github.com/danielgraf/boat-brain-audio-alerter/internal/config"
	"github.com/danielgraf/boat-brain-audio-alerter/internal/heading"
	"github.com/danielgraf/boat-brain-audio-alerter/internal/pid"
	"github.com/danielgraf/boat-brain-audio-alerter/internal/ram"
)

// Mode is the steering reference.
type Mode string

const (
	HeadingHold Mode = "heading"
	TrackCOG    Mode = "cog"
)

// Tick is the diagnostics returned from each Update call.
type Tick struct {
	Engaged         bool    `json:"engaged"`
	Mode            Mode    `json:"mode"`
	DesiredHeading  float64 `json:"desired_heading"`
	DesiredCOG      float64 `json:"desired_cog"`
	HeadingSetpoint float64 `json:"heading_setpoint"`
	MeasuredHeading float64 `json:"measured_heading"`
	MeasuredCOG     float64 `json:"measured_cog"`
	Error           float64 `json:"error"`
	YawRate         float64 `json:"yaw_rate"`
	Crab            float64 `json:"crab"`
	RudderCommand   float64 `json:"rudder_command"`
	RudderAchieved  float64 `json:"rudder_achieved"`
	SpeedKnots      float64 `json:"speed_knots"`
	Reason          string  `json:"reason"`
	OffCourse       bool    `json:"off_course"` // heading off the locked course beyond alarm limits
}

// CourseKeeper is the closed-loop heading/COG controller.
type CourseKeeper struct {
	cfg config.Autopilot
	ram ram.Actuator
	pid *pid.PID

	engaged        bool
	mode           Mode
	desiredHeading float64
	desiredCOG     float64

	headingSetpoint float64
	haveSetpoint    bool

	prevHeading     float64
	havePrev        bool
	filteredHeading float64
	haveFiltered    bool
	filteredCOG     float64
	haveFilteredCOG bool

	crabIntegral float64
	trackInit    bool
	baseGains    pid.Gains

	offCourseFor float64 // seconds the heading has been beyond the alarm limit
	offCourse    bool
}

// New builds a CourseKeeper around a config and an actuator.
func New(cfg config.Autopilot, actuator ram.Actuator) *CourseKeeper {
	return &CourseKeeper{
		cfg:       cfg,
		ram:       actuator,
		pid:       pid.New(cfg.PID),
		mode:      HeadingHold,
		baseGains: cfg.PID.Gains,
	}
}

// PID exposes the inner controller (for the visualizer's tuning readouts).
func (k *CourseKeeper) PID() *pid.PID { return k.pid }

// SetBaseGains updates the un-scheduled gains (e.g. after a live re-tune). The
// speed scheduler rescales from these each tick.
func (k *CourseKeeper) SetBaseGains(g pid.Gains) {
	k.baseGains = g
	k.pid.Config.Gains = g
}

// Engaged reports whether the pilot is driving the helm.
func (k *CourseKeeper) Engaged() bool { return k.engaged }

// Mode returns the current steering reference.
func (k *CourseKeeper) Mode() Mode { return k.mode }

// DesiredHeading / DesiredCOG expose the current setpoints.
func (k *CourseKeeper) DesiredHeading() float64 { return k.desiredHeading }
func (k *CourseKeeper) DesiredCOG() float64     { return k.desiredCOG }

// Engage engages the pilot. With no explicit setpoint it locks onto the current
// heading ("press AUTO and hold what I've got").
func (k *CourseKeeper) Engage(currentHeading float64) {
	k.desiredHeading = heading.Normalize360(currentHeading)
	k.headingSetpoint = k.desiredHeading
	k.haveSetpoint = true
	k.engaged = true
	k.trackInit = false
	k.crabIntegral = 0
	k.offCourseFor = 0
	k.offCourse = false
	k.pid.Reset()
	k.prevHeading = heading.Normalize360(currentHeading)
	k.havePrev = true
	k.filteredHeading = heading.Normalize360(currentHeading)
	k.haveFiltered = true
}

// Standby disengages: stop driving the ram and let the helm go free.
func (k *CourseKeeper) Standby() {
	k.engaged = false
	k.ram.Stop()
}

// SetMode switches steering reference. Switching to COG-track re-learns the crab
// angle bumplessly on the next usable fix.
func (k *CourseKeeper) SetMode(m Mode) {
	if m != k.mode {
		k.mode = m
		k.trackInit = false
	}
}

// SetHeading dials in a new heading to hold (deg).
func (k *CourseKeeper) SetHeading(h float64) { k.desiredHeading = heading.Normalize360(h) }

// SetCOG dials in a new course over ground to hold (deg).
func (k *CourseKeeper) SetCOG(c float64) { k.desiredCOG = heading.Normalize360(c) }

// AdjustHeading nudges the heading setpoint (e.g. ±1 / ±10 buttons).
func (k *CourseKeeper) AdjustHeading(delta float64) {
	k.desiredHeading = heading.Normalize360(k.desiredHeading + delta)
}

// AdjustCOG nudges the COG setpoint.
func (k *CourseKeeper) AdjustCOG(delta float64) {
	k.desiredCOG = heading.Normalize360(k.desiredCOG + delta)
}

// lowpassAngle moves a filtered angle toward a target along the shortest arc.
func lowpassAngle(filtered, target, tau, dt float64) float64 {
	if tau <= 0 {
		return target
	}
	alpha := dt / (tau + dt)
	return heading.Normalize360(filtered + alpha*heading.Error(target, filtered))
}

func (k *CourseKeeper) applySpeedScheduling(speed float64) {
	if !k.cfg.SpeedScheduling {
		k.pid.Config.Gains = k.baseGains
		return
	}
	v := math.Max(speed, k.cfg.MinSpeedKnots)
	scaled := k.cfg.Nomoto.ScaledToSpeed(v, k.cfg.RefSpeedKnots)
	if g, err := k.cfg.GainsFor(scaled); err == nil {
		k.pid.Config.Gains = g
	} else {
		k.pid.Config.Gains = k.baseGains
	}
}

// outerLoop updates the heading setpoint from the COG error (cascade).
func (k *CourseKeeper) outerLoop(cog, measuredHeading, dt float64) float64 {
	o := k.cfg.Outer
	if k.haveFilteredCOG {
		k.filteredCOG = lowpassAngle(k.filteredCOG, cog, o.COGFilterTau, dt)
	} else {
		k.filteredCOG = heading.Normalize360(cog)
		k.haveFilteredCOG = true
	}
	e := heading.Error(k.desiredCOG, k.filteredCOG)
	nearCourse := math.Abs(e) <= o.ActiveBand

	switch {
	case !k.trackInit:
		// Seed the crab integral with the crab the boat is currently flying
		// (heading - desiredCOG) so acquisition steers from a sensible offset
		// rather than lurching, and the integral already holds ~the value it
		// needs to learn.
		k.crabIntegral = clamp(heading.Error(measuredHeading, k.desiredCOG), -o.MaxCrab, o.MaxCrab)
		k.trackInit = true
	case nearCourse:
		// Integrate only near course, with anti-windup (don't push a saturated
		// crab further into its limit). Far from course the error is large and
		// one-signed for many seconds, which would wind the crab up the wrong way.
		cand := clamp(k.crabIntegral+o.KiCrab*e*dt, -o.MaxCrab, o.MaxCrab)
		if (cand < o.MaxCrab || e < 0) && (cand > -o.MaxCrab || e > 0) {
			k.crabIntegral = cand
		}
	}

	// Near course, add proportional fine-tuning; during acquisition steer on the
	// learned crab alone so the big COG error doesn't fling the setpoint past
	// target.
	crab := k.crabIntegral
	if nearCourse {
		crab = o.KpCrab*e + k.crabIntegral
	}
	crab = clamp(crab, -o.MaxCrab, o.MaxCrab)
	return heading.Normalize360(k.desiredCOG + crab)
}

// Update runs one control tick.
//
// measuredHeading is the compass heading (deg). dt is seconds since the last
// tick. sensorYawRate, speedKnots, cog and sog are optional (pass nil when the
// value isn't available); cog/sog drive TrackCOG mode.
func (k *CourseKeeper) Update(measuredHeading, dt float64, sensorYawRate, speedKnots, cog, sog *float64) Tick {
	raw := heading.Normalize360(measuredHeading)
	speed := k.cfg.RefSpeedKnots
	if speedKnots != nil {
		speed = *speedKnots
	}

	// Low-pass the compass input (wrap-aware) so the derivative term isn't fed
	// noise. Real fluxgate/GPS compasses are physically damped; this models it.
	if k.cfg.HeadingFilterTau > 0 && k.haveFiltered {
		k.filteredHeading = lowpassAngle(k.filteredHeading, raw, k.cfg.HeadingFilterTau, dt)
	} else {
		k.filteredHeading = raw
		k.haveFiltered = true
	}
	measured := k.filteredHeading

	var rate float64
	switch {
	case sensorYawRate != nil:
		rate = *sensorYawRate
	case k.havePrev:
		rate = heading.YawRate(k.prevHeading, measured, dt)
	}
	k.prevHeading = measured
	k.havePrev = true

	measuredCOG := 0.0
	if cog != nil {
		measuredCOG = heading.Normalize360(*cog)
	}

	if !k.engaged {
		k.offCourseFor, k.offCourse = 0, false
		return Tick{
			Mode: k.mode, DesiredHeading: k.desiredHeading, DesiredCOG: k.desiredCOG,
			MeasuredHeading: measured, MeasuredCOG: measuredCOG, YawRate: rate,
			RudderCommand: k.ram.RudderAngle(), RudderAchieved: k.ram.RudderAngle(),
			SpeedKnots: speed, Reason: "standby",
		}
	}

	// Decide the heading setpoint.
	reason := "hold"
	cogUsable := k.mode == TrackCOG && cog != nil && sog != nil && *sog >= k.cfg.Outer.MinTrackSpeed
	switch {
	case cogUsable:
		k.headingSetpoint = k.outerLoop(*cog, measured, dt)
		reason = "track-cog"
	case k.mode == TrackCOG:
		// COG unusable (too slow / no fix): hold the last heading setpoint.
		k.trackInit = false
		reason = "cog-lowspeed-hold"
	default:
		k.headingSetpoint = k.desiredHeading
	}
	if !k.haveSetpoint {
		k.headingSetpoint = k.desiredHeading
		k.haveSetpoint = true
	}

	errDeg := heading.Error(k.headingSetpoint, measured)

	// Off-course alarm: heading off the locked course beyond the limit for too
	// long (ST4000 default 20° for 20 s).
	if k.cfg.OffCourseDeg > 0 && math.Abs(errDeg) > k.cfg.OffCourseDeg {
		k.offCourseFor += dt
	} else {
		k.offCourseFor = 0
	}
	k.offCourse = k.cfg.OffCourseSecs > 0 && k.offCourseFor >= k.cfg.OffCourseSecs

	// Displayed crab: the actual offset of the bow from the ground track.
	crab := heading.Error(measured, measuredCOG)
	if k.mode != TrackCOG {
		crab = 0
	}

	// Low-steerage guard: no water flow over the rudder means no authority, so
	// hold station rather than winding up and slamming the helm over.
	if speedKnots != nil && *speedKnots < k.cfg.MinSpeedKnots {
		achieved := k.ram.Command(k.ram.RudderAngle(), dt)
		return Tick{
			Engaged: true, Mode: k.mode, DesiredHeading: k.desiredHeading, DesiredCOG: k.desiredCOG,
			HeadingSetpoint: k.headingSetpoint, MeasuredHeading: measured, MeasuredCOG: measuredCOG,
			Error: errDeg, YawRate: rate, Crab: crab, RudderCommand: k.ram.RudderAngle(),
			RudderAchieved: achieved, SpeedKnots: speed, Reason: "low-steerage", OffCourse: k.offCourse,
		}
	}

	k.applySpeedScheduling(speed)
	rudderCmd := k.pid.Update(errDeg, rate, dt)
	achieved := k.ram.Command(rudderCmd, dt)

	return Tick{
		Engaged: true, Mode: k.mode, DesiredHeading: k.desiredHeading, DesiredCOG: k.desiredCOG,
		HeadingSetpoint: k.headingSetpoint, MeasuredHeading: measured, MeasuredCOG: measuredCOG,
		Error: errDeg, YawRate: rate, Crab: crab, RudderCommand: rudderCmd,
		RudderAchieved: achieved, SpeedKnots: speed, Reason: reason, OffCourse: k.offCourse,
	}
}

func clamp(x, lo, hi float64) float64 {
	if x < lo {
		return lo
	}
	if x > hi {
		return hi
	}
	return x
}
